C/common/logger.cpp

```diff
@@ -98,6 +102,73 @@ void Logger::setMinLevel(const string& level)
 	}
 }
 
+bool Logger::registerInterceptor(LogLevel level, LogInterceptor callback, void* userData)
+{
+	std::lock_guard<std::mutex> lock(m_interceptorMapMutex);
+	auto newMap = std::make_shared<InterceptorMap>(*m_interceptors); // Copy current map
+	auto it = newMap->emplace(level, LogInterceptorNode{callback, userData}); // Add new interceptor
+	if (it != newMap->end())
+	{
+		m_interceptors = newMap; // Update the shared_ptr
+		return true;
+	}
+	return false;
+}
+
+bool Logger::unregisterInterceptor(LogLevel level, LogInterceptor callback)
+{
+	std::lock_guard<std::mutex> lock(m_interceptorMapMutex);
+	auto newMap = std::make_shared<InterceptorMap>(*m_interceptors); // Copy current map
+	auto range = newMap->equal_range(level);
+	for (auto it = range.first; it != range.second; ++it)
+	{
+		if (it->second.callback == callback)
+		{
+			newMap->erase(it); // Remove the interceptor
+			m_interceptors = newMap; // Update the shared_ptr
+			return true; // Remove only one matching entry
+		}
+	}
+	return false;
+}
+
+void Logger::executeInterceptor(LogLevel level, const std::string& message)
+{
+	std::shared_ptr<InterceptorMap> interceptors;
+	{
+		std::lock_guard<std::mutex> lock(m_interceptorMapMutex);
+		interceptors = m_interceptors; // Get a consistent snapshot of the interceptor map
+	}
+
+	auto range = interceptors->equal_range(level);
+	for (auto it = range.first; it != range.second; ++it)
+	{
+		// Offload the callback execution to a separate thread using std::async
+		m_futures.emplace_back(std::async(std::launch::async, [it, level, message]() {
```

Comment:
std::async is a bad choice here as it is really best used for expensive operations, it can lead to many threads being created and requires the cleanup of the futures. It would be ,uh better to have a queue with a single thread to process it. This would also have the advantage of preserving the order or intercepted log messages.